File: JS/Tree/236. Lowest Common Ancestor of a Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    const stack = [[root, []]];
    let pPath, qPath;
    while(stack.length && (!pPath || !qPath)) {
        const [node, ancestor] = stack.shift();
        const newAncestor = ancestor.concat([node]);
        if (node.val === p.val) {
            pPath = newAncestor;
        }
        if (node.val === q.val) {
            qPath = newAncestor;
        }
        if (node.left) {
            stack.push([node.left, newAncestor]);
        }
        if (node.right) {
            stack.push([node.right, newAncestor]);
        }
    }
    let result = root;
    while(pPath.length && qPath.length) {
        const x = pPath.shift();
        const y = qPath.shift();
        if (x.val === y.val) {
            result = x;
        } else {
            break;
        }
    }
    return result;
};

File: Golang/Tree/236.lowest-common-ancestor-of-a-binary-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestorRecursive(root, p, q *TreeNode) *TreeNode {
	if root == nil || root == p || root == q {
		return root
	}
	left := lowestCommonAncestorRecursive(root.Left, p, q)
	right := lowestCommonAncestorRecursive(root.Right, p, q)
	if left != nil && right != nil {
		// it means for given root, we can find p and q in it's left/right sub trees
		return root
	} else if left == nil {
		return right
	}
	return left
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	type ComplextNode struct {
		Node      *TreeNode
		Ancestors []*TreeNode
	}
	var getAncestors func(root *TreeNode, node *TreeNode) []*TreeNode
	getAncestors = func(root *TreeNode, node *TreeNode) []*TreeNode {
		var ret []*TreeNode
		stack := []ComplextNode{ComplextNode{
			Node:      root,
			Ancestors: []*TreeNode{},
		}}
		for len(stack) > 0 {
			children := []ComplextNode{}
			for i := 0; i < len(stack); i++ {
				temp := stack[i]
				if temp.Node == node {
					ret = append(temp.Ancestors, temp.Node)
					break
				}
				if temp.Node.Left != nil {
					children = append(children, ComplextNode{
						Node:      temp.Node.Left,
						Ancestors: append(temp.Ancestors, temp.Node),
					})
				}
				if temp.Node.Right != nil {
					children = append(children, ComplextNode{
						Node:      temp.Node.Right,
						Ancestors: append(temp.Ancestors, temp.Node),
					})
				}
			}
			stack = children
		}
		return ret
	}
	ancestorsP := getAncestors(root, p)
	ancestorsQ := getAncestors(root, q)
	var ret *TreeNode
	i := 0
	for i < len(ancestorsP) && i < len(ancestorsQ) {
		if ancestorsP[i] == ancestorsQ[i] {
			ret = ancestorsP[i]
		} else {
			break
		}
		i++
	}
	return ret
}
